applications/jonny5/src/j5_balance_crawler.erl

```diff
@@ -0,0 +1,242 @@
+    end.
+
+%%====================================================================
+%% gen_server callbacks
+%%====================================================================
+
+%%--------------------------------------------------------------------
+%% Function: init(Args) -> {ok, State} |
+%%                         {ok, State, Timeout} |
+%%                         ignore               |
+%%                         {stop, Reason}
+%% Description: Initiates the server
+%%--------------------------------------------------------------------
+init([]) ->
+    wh_util:put_callid(?MODULE),
+    sleep_and_cast(0,'crawl_accounts'),
+    {'ok', []}.
+
+%%--------------------------------------------------------------------
+%% Function: %% handle_call(Request, From, State) -> {reply, Reply, State} |
+%%                                      {reply, Reply, State, Timeout} |
+%%                                      {noreply, State} |
+%%                                      {noreply, State, Timeout} |
+%%                                      {stop, Reason, Reply, State} |
+%%                                      {stop, Reason, State}
+%% Description: Handling call messages
+%%--------------------------------------------------------------------
+handle_call(_Request, _From, State) ->
+    Reply = {'error', 'not_implemented'},
+    {'reply', Reply, State}.
+
+%%--------------------------------------------------------------------
+%% Function: handle_cast(Msg, State) -> {noreply, State} |
+%%                                      {noreply, State, Timeout} |
+%%                                      {stop, Reason, State}
+%% Description: Handling cast messages
+%%--------------------------------------------------------------------
+handle_cast('crawl_accounts', []) ->
+    Cycle = whapps_config:get_integer(?APP_NAME, <<"balance_crawler_cycle_ms">>, ?DEFAULT_CRAWLER_CYCLE),
+    spawn_sleep_and_cast(Cycle, 'crawl_accounts'),
+    case j5_channels:accounts() of
+        [] ->
+            {'noreply', [], 'hibernate'};
+        Accounts ->
+            lager:debug("check ~p account(s) for zero balance", [length(Accounts)]),
+            sleep_and_cast(0, 'next_account'),
+            {'noreply', Accounts}
+    end;
+
+%%--------------------------------------------------------------------------------------------
+%% If crawler still checking account from previous cycle, warn and try again after some delay
+%%--------------------------------------------------------------------------------------------
+handle_cast('crawl_accounts', [_|_] = Accounts) ->
+    lager:warning("delay next crawler cycle for ~p ms, ~p account(s) to go. Please tune crawler settings", [?OVERLOAD_DELAY, length(Accounts)]),
+    spawn_sleep_and_cast(?OVERLOAD_DELAY, 'crawl_accounts'),
+    {'noreply', Accounts};
+
+handle_cast('next_account', []) ->
+    {'noreply', [], 'hibernate'};
+
+handle_cast('next_account', [Account|Accounts]) ->
+    maybe_disconnect_account(Account),
```

Comment:
I think we could get a better separation of concerns here.

**gen_server**

1. gen_server tracks the timers for checking accounts for per-minute depletion
2. gen_server spawns worker on timer expiration to check accounts
3. gen_server waits to start next timer until worker exits

**worker**
1. fetches list of accounts
2. maybe disconnects each account

I think this could let us extend the worker to disconnect calls that have used up allotments and possibly other strategies. If we put this worker code in a separate module, we could also expose some of the functionality via `jonny5_maintenance` and get some nice SUP commands to inspect and disconnect accounts as well.

Thoughts?